### DISCUSSION TITLE
Android strongswan reconnect on failure

### DISCUSSION THREAD

**Discussion by exoticprgmr** [strongswan/strongswan#2027] [Q&A] [answered]
Good afternoon,

First and foremost, I would like to express my gratitude to the StrongSwan developers, particularly Tobias Brunner, whose explanations on different forums were instrumental in helping me set up a working VPN connection. I've successfully established an "almost ideal" Always-on VPN connection between my rented VPS server (Debian 12 - StrongSwan 5.9.8) and my Android 14 client (Samsung A53 5G).

The primary goal was to have a fault-tolerant connection, ensuring that no single packet can go to the internet unencrypted. I've disabled reauth and rekey on the server, and the IKEv2 native client on Android is used to enable a VPN connection straight after phone restarts.

Everything works well until the Android phone loses connection, either due to a poor signal on WiFi or 4G. The issue arises when the Android client, after losing connection, sends Dead Peer Detection (DPD) requests every few minutes. If these requests don't receive a reply and reach six retransmissions on port 4500, the client attempts to re-negotiate IPSec SA's, falling back to port 500. However, if this fails, the VPN notification disappears, and the connection stalls until manually reconnected in the Settings.

I expected the Android client to automatically re-connect to the VPN profile if the connection fails for any reason, but unfortunately, this is not happening.

A point of interest is MOBIKE (RFC 4555 extensions), which is currently supported on the client and disabled on the server. Temporarily enabling it on the server prevents DPD from being sent during the connection. Strangely, after the connection is established, the client cannot reach the server's IP address, and I receive the message: "no route found to reach MY.EXT.IP.ADDR, MOBIKE update deferred.". Connection works perfectly fine until client attempts first CHILD_SA rekey, then I get "no route found to reach peer, path probing deferred.".

I've also noticed that if I temporarily enable re-keying on the server, there's an issue if the client is not reachable, and the server clears the connection. The client gets stalled if it comes back online. This is definitely what I don't want to happen.

Has anyone else experienced this behavior on Android, and is there a way to prompt the internal VPN connection to restart if the connection stalls? The server configuration seems perfect; it's just the client that sends DPDs, fails on poor connection, and doesn't restart as a result. Unfortunately there is also not much of StrongSwan configuration on Android.

I appreciate any insights or suggestions on resolving this issue.

Environment:

    Server: Debian 12, StrongSwan 5.9.8
    Client: Android 14, Samsung A53 5G (charon - version unknown)
    
ipsec.conf (server):
config setup
     charondebug="ike 1, knl 1, cfg 0"
     uniqueids=yes

conn %default
      compress=no
      type=tunnel
      keyexchange=ikev2
      fragmentation=yes
      forceencaps=yes
      rekey=no
      reauth=no
      right=%any
      left=MY.EXT.IP.ADDR
      leftsubnet=0.0.0.0/0
      leftauth=pubkey
      leftid=@vpn.myfqdn.net
      leftcert=serverCert.der
      leftsendcert=always
      ike=aes256-sha384-prfsha384-ecp384!
      esp=aes256gcm16,aes256gcm16-ecp384!
      mobike=no

conn ipsec-ikev2-vpn
      auto=add
      ikelifetime=24h
      lifetime=12h
      rightsourceip=%dhcp
      rightauth=eap-mschapv2
      rightsendcert=never
      eap_identity=%identity
      dpddelay=0

**Comment by tobiasbrunner** [comment, marked as answer]
> I expected the Android client to automatically re-connect to the VPN profile if the connection fails for any reason, but unfortunately, this is not happening.

That's apparently not the case. I'd recommend using [our Android app](https://docs.strongswan.org/docs/5.9/os/androidVpnClient.html) instead, which tries to reconnect until the user manually disconnects the VPN should it ever be unable to reach the server (it does not actively send DPDs, though, so it only will run into this if it rekeys the SAs).

> Strangely, after the connection is established, the client cannot reach the server's IP address, and I receive the message: "no route found to reach MY.EXT.IP.ADDR, MOBIKE update deferred.".

That sounds weird, but might just be an issue in Samsung's VPN client (which is based on strongSwan, but what they do exactly with routing etc. I don't know). MOBIKE should be enabled when using the strongSwan app.

> I've also noticed that if I temporarily enable re-keying on the server, there's an issue if the client is not reachable, and the server clears the connection. 

Yes, that can be an issue with mobile clients. It could be avoided with retransmission and lifetime settings that allow enough time/retries for the client to be unavailable for longer stretches of time. And depending on the client, it might just recreate the connection if the server removed the state. But disabling actively sending IKE packets from the server might be more stable.

> Has anyone else experienced this behavior on Android, and is there a way to prompt the internal VPN connection to restart if the connection stalls?

Note that Samsung does currently not use Google's IKEv2 VPN client (that's built into Android 11+), but uses their own client based on strongSwan (heavily patched). So it can't directly be compared to Android.

**Reply by exoticprgmr** [reply]
I've started using strongSwan app on Android, and it has provided me with a remarkably stable connection to my VPN server. I've consistently achieved 60-hour uptime, thanks to the MOBIKE extensions. This level of reliability is unparalleled among other VPN applications I've tried on the Android platform.

One minor drawback I've encountered is that packets remain unencrypted until I unlock my phone after restart. However, I understand that this limitation is inherent to the Android system, and it doesn't seem to be within your control. Fortunately, strongSwan promptly ensures that no packets go unencrypted once phone is unlocked and connection is reestablished after any interruption, providing a robust solution to this challenge.

Despite the occasional inconvenience, I find that the overall performance and stability of strongSwan on Android surpass that of other VPN apps. Your guidance on retransmission and lifetime settings to handle potential issues with mobile clients has been particularly insightful and I adjusted my config accordingly.

Once again, thank you for your dedication to this project.

I'm marking your reponse as an answer.

**Reply by tobiasbrunner** [reply]
Thanks for the feedback.

> One minor drawback I've encountered is that packets remain unencrypted until I unlock my phone after restart.

Have you enabled the "Always-on VPN" option in the Android VPN settings for the strongSwan app? And then the "Block connection without VPN" option? Both are system settings that can't be controlled by the app itself, so you have to do that in the system's VPN settings.

**Reply by exoticprgmr** [reply]
Thank you for your suggestion. Unfortunately, the "Block connection without VPN" option won't work for me because in such case I can't reach the phone either directly by IP or via VPN after device restart but before it is unlocked. For example, if I lose my device and it is restarted, there is no way to find it remotely with the option ticked in.

Currently, I have the always-on VPN enabled but no "Block connection without VPN" option checked. This way, I can find the phone using "Samsung SmartThings Find" before the phone is unlocked after a restart. Once the phone is unlocked, the Strongswan VPN for Android starts and protects all traffic.

In ideal world, my device should be VPN protected straight after reboot, but I understand it's an Android OS limitation.
